### ISSUE TITLE
Exception handling in transform or safeParse

### ISSUE THREAD

**Issue by zaripych** [colinhacks/zod#3077]
```
import { z } from "zod";

const invalidJson = '{"ax: 1}'

console.log(
  z
    .string()
    .transform((val) => JSON.parse(val))
    .safeParse(invalidJson)
);
```

--

Expected:
```json
{ "success": false, "error": ... }
```

--

Result:
It throws the exception. "Unexpected end of JSON input".

--

Basically exceptions are not handled in `transform` or `safeParse` as you would expect them to. 

I've checked the documentation and it doesn't mention anything about handling exceptions in `transform`.

There seem to be similar behaviour for `refine` as well.

zod package version - 3.22.4

**Comment by JacobWeisenburger** [comment]
Perhaps this will help:
https://github.com/JacobWeisenburger/zod_utilz#json

**Comment by zaripych** [comment]
Yeah, I ended up handling exception on my own. But it felt surprising that `safeParse` actually doesn't handle exceptions. 👍

**Comment by JacobWeisenburger** [comment]
are you wanting this to stay open as an enhancement? or are you satisfied with your solution?

**Comment by zaripych** [comment]
Thank you. Yes. It seems reasonable to me to handle the exception in `safeParse` at the very least or improve the `.transform` to handle the exceptions and add the error to the context automatically, but if that complicates the use of the library or makes breaking changes that people don't want - I'm fine with closing it.

I think the main benefit of handling the exception inside the `transform` by users is more fine grained control over the error message. But if that is important to users they can always add it later.

The benefit of handling the exception in Zod itself is brewety and simplicity. Ie:

```
.tranform((value) => JSON.parse(value))
```

is shorter than

```
.tranform((value, ctx) => {
  try {
    return JSON.parse(value)
  } catch (err) {
    ctx.addIssue('custom', ...);
    return z.NEVER;
  }
})
```

There is just so much going on in `.transform` now ... `ctx`, `addIssue`, `z.NEVER` ... many things to be done "correctly".

Does this seem like a good direction to you?

**Comment by mmkal** [comment]
I agree, this is a common enough use-case that I think it makes sense for zod to handle correctly. Otherwise, the path of least resistance is to do `z.object({ fooConfig: z.string().transform(s => JSON.parse(s)).pipe(z.object({ abc: z.number() }))`

That's what we have in our codebase now in a few places, and it's hard to justify a code review comment saying "turn that simple, easy to understand `.transform` into an eleven-line function, featuring a try-catch, a mysterious `ctx` variable that most devs aren't familiar with, and the `z.ZodIssueCode` enum.

Instead I'd propose one or two new zod features:

First, a first-class JSON helper on string, since JSON-parsing is so common, e.g. `z.string().json(z.object({ foo: z.number(() }))` - this is just sugar for:

```ts
z.string()
  .transform((str, ctx) => {
    try {
      JSON.parse(str)
    } catch (error) {
      ctx.addIssue({
        code: z.ZodIssueCode.invalid_json, // new error code
        message: (error as Error).message,
      })
      return z.NEVER
  })
  .pipe(z.object({ foo: z.number() })
```

Second, for other parsing methods than JSON, a `.safeTransform` method on ZodType, e.g. `z.string().safeTransform(jsYaml.load)`. This would be sugar for a try-catched `.transform`:

```ts
z.string()
  .transform((str, ctx) => {
    try {
      jsYaml.load(str)
    } catch (error) {
      ctx.addIssue({
        code: z.ZodIssueCode.custom,
        message: (error as Error).message,
      })
      return z.NEVER
  })
```

These two could be proposed in separate pull requests in case one is more controversial than the other.

What do you think @colinhacks @JacobWeisenburger? I'd be happy to open a PR for both.

**Comment by mmkal** [comment]
I added json support in https://github.com/colinhacks/zod/pull/3109

But looks like it's now removed/tests got commented out: 
https://github.com/colinhacks/zod/blob/main/packages/zod/src/v4/classic/tests/json.test.ts

@colinhacks was there something you wanted to change about the api/implementation or did you decide against it in the end?
